In [1]:
!pip install rapidfuzz

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 26.0 MB/s eta 0:00:00


In [2]:
!pip install -q deepeval

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 23.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 598.8/598.8 kB 27.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.4/69.4 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.6/129.6 kB 9.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 53.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.4/46.4 kB 1.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.7/40.7 kB 2.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
huggingface-hub 1.28.0 requires click<9.0.0,>=8.4.2, but you have click 8.3.3 which is incompatible.


In [3]:
!pip install -q "ragas==0.3.9" "datasets" "langchain-google-genai"
!pip install -q "langchain-text-splitters" "pypdf" "pandas" "sentence-transformers" "tiktoken"
!pip install -q "bm25s" "chromadb"
!pip uninstall -y -q transformers tokenizers FlagEmbedding
!pip install -U -q "transformers>=4.46,<5" "tokenizers>=0.20.3" "FlagEmbedding"

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.5/56.5 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 366.7/366.7 kB 16.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.6/81.6 kB 8.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 49.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 571.5/571.5 kB 48.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 222.1/222.1 kB 23.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 260.1/260.1 kB 27.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 97.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.1/73.1 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 125.1/125.1 kB 11.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 96.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 125.3/

In [4]:
!pip show ragas

Name: ragas
Version: 0.3.9
Summary: Evaluation framework for RAG and LLM applications
Home-page: https://github.com/explodinggradients/ragas
Author: 
Author-email: 
License: Apache License
                           Version 2.0, January 2004
                        http://www.apache.org/licenses/

   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION

   1. Definitions.

      "License" shall mean the terms and conditions for use, reproduction,
      and distribution as defined by Sections 1 through 9 of this document.

      "Licensor" shall mean the copyright owner or entity authorized by
      the copyright owner that is granting the License.

      "Legal Entity" shall mean the union of the acting entity and all
      other entities that control, are controlled by, or are under common
      control with that entity. For the purposes of this definition,
      "control" means (i) the power, direct or indirect, to cause the
      direction or management of such entity, wheth

In [5]:
import sys
import types

# ragas/llms/base.py unconditionally does:
#   from langchain_community.chat_models.vertexai import ChatVertexAI
#   from langchain_community.llms import VertexAI
# Newer langchain_community removed both. We don't use VertexAI (we use
# Gemini via langchain-google-genai instead), so we inject harmless stub
# modules just so ragas's import line succeeds.

import langchain_community.chat_models as _lcc_chat_models

class _StubChatVertexAI:
    def __init__(self, *args, **kwargs):
        raise NotImplementedError("VertexAI stub — this project doesn't use VertexAI.")

_vertexai_stub_module = types.ModuleType("langchain_community.chat_models.vertexai")
_vertexai_stub_module.ChatVertexAI = _StubChatVertexAI
sys.modules["langchain_community.chat_models.vertexai"] = _vertexai_stub_module
_lcc_chat_models.vertexai = _vertexai_stub_module

import langchain_community.llms as _lcc_llms
if not hasattr(_lcc_llms, "VertexAI"):
    class _StubVertexAI:
        def __init__(self, *args, **kwargs):
            raise NotImplementedError("VertexAI stub — this project doesn't use VertexAI.")
    _lcc_llms.VertexAI = _StubVertexAI

print("VertexAI stubs installed — ragas import should now succeed.")

/tmp/ipykernel_689/2880259009.py:11: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  import langchain_community.chat_models as _lcc_chat_models


VertexAI stubs installed — ragas import should now succeed.


In [6]:
import os
import math
import json
from pathlib import Path
from collections import defaultdict

import numpy as np
import pandas as pd
import bm25s
import chromadb
from sentence_transformers import SentenceTransformer
from FlagEmbedding import FlagReranker

from datasets import Dataset
from ragas import evaluate
from ragas.metrics import faithfulness, answer_relevancy, context_precision, context_recall
from ragas.llms import LangchainLLMWrapper
from ragas.embeddings import LangchainEmbeddingsWrapper
from langchain_google_genai import ChatGoogleGenerativeAI, GoogleGenerativeAIEmbeddings

from google import genai

In [7]:
import os, requests

os.makedirs("sample_docs", exist_ok=True)

FILES = {
    "sample_docs/nist_800-61r2.pdf": "https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-61r2.pdf",
    "sample_docs/attention_is_all_you_need.pdf": "https://arxiv.org/pdf/1706.03762",
    "sample_docs/langchain_readme.md": "https://raw.githubusercontent.com/langchain-ai/langchain/master/README.md",
    "sample_docs/pride_and_prejudice.txt": "https://www.gutenberg.org/cache/epub/1342/pg1342.txt",
}

headers = {"User-Agent": "Mozilla/5.0 (research; contact: rooofeee21@gmail.com)"}  # SEC requires a UA header
for path, url in FILES.items():
    r = requests.get(url, headers=headers, timeout=30)
    r.raise_for_status()
    with open(path, "wb") as f:
        f.write(r.content)
    print(f"downloaded {path} ({len(r.content):,} bytes)")

downloaded sample_docs/nist_800-61r2.pdf (1,744,473 bytes)
downloaded sample_docs/attention_is_all_you_need.pdf (2,215,244 bytes)
downloaded sample_docs/langchain_readme.md (5,933 bytes)
downloaded sample_docs/pride_and_prejudice.txt (772,386 bytes)


In [8]:
"""
chunking_harness.py
=====================
Runs all 7 chunking strategies against one or more documents (PDF/TXT/MD/HTML)
at multiple target chunk sizes, and produces a comparison table of structural
quality metrics.

Metrics per (document, method, chunk size):
    chunk_count            - how many chunks were produced
    avg_len / median_len   - chunk length in characters
    min_len / max_len
    std_len                - length variance (tighter is usually "cleaner")
    pct_mid_sentence_cut   - % of chunks NOT ending on sentence punctuation
                              (a structural-quality proxy — not a retrieval
                              quality measure; see note at bottom of script)
    elapsed_sec            - processing time

Note: markdown_aware only runs on .md files, html_aware only on .html/.htm
files (they need that structure to split on). The other 5 strategies run on
any text-bearing file (.pdf, .txt, .md).

Usage:
    python chunking_harness.py doc1.pdf doc2.md doc3.html
    python chunking_harness.py --dir ./my_pdfs
    python chunking_harness.py --dir ./my_pdfs --sizes 256 512 --out results.csv

Install:
    pip install langchain-text-splitters pypdf pandas sentence-transformers tiktoken
    (tiktoken and sentence-transformers are optional — the script degrades
    gracefully without them, see the try/except blocks below)
"""

import re
import time
from pathlib import Path

import pandas as pd

from langchain_text_splitters import (
    CharacterTextSplitter,
    RecursiveCharacterTextSplitter,
    MarkdownHeaderTextSplitter,
    HTMLHeaderTextSplitter,
)

# ---------------------------------------------------------------------------
# Optional dependencies — script still runs (with reduced coverage) without them
# ---------------------------------------------------------------------------
try:
    import tiktoken
    _ENC = tiktoken.get_encoding("cl100k_base")

    def count_tokens(text):
        return len(_ENC.encode(text))
except ImportError:
    def count_tokens(text):
        return len(text.split())  # rough word-count fallback


def tokens_to_chars(n_tokens):
    """~4 chars/token is a standard rough ratio for English prose."""
    return int(n_tokens * 4)


try:
    from sentence_transformers import SentenceTransformer
    import numpy as np

    _MODEL = None

    def get_model():
        global _MODEL
        if _MODEL is None:
            _MODEL = SentenceTransformer("all-MiniLM-L6-v2")
        return _MODEL

    SEMANTIC_AVAILABLE = True
except ImportError:
    SEMANTIC_AVAILABLE = False

try:
    from pypdf import PdfReader

    PDF_AVAILABLE = True
except ImportError:
    PDF_AVAILABLE = False


# ---------------------------------------------------------------------------
# Text extraction
# ---------------------------------------------------------------------------

def extract_text(path: Path) -> list[dict]:
    """Returns a list of {"page": int, "text": str} — one entry per PDF page,
    or a single entry (page=1) for non-paginated formats."""
    suffix = path.suffix.lower()
    if suffix == ".pdf":
        if not PDF_AVAILABLE:
            raise RuntimeError("pip install pypdf to read PDFs")
        reader = PdfReader(str(path))
        return [
            {"page": i + 1, "text": (p.extract_text() or "").strip()}
            for i, p in enumerate(reader.pages)
        ]
    if suffix in (".htm", ".html"):
        from bs4 import BeautifulSoup
        raw = path.read_text(encoding="utf-8", errors="ignore")
        soup = BeautifulSoup(raw, "html.parser")
        for tag in soup(["script", "style"]):
            tag.decompose()
        return [{"page": 1, "text": soup.get_text(separator="\n")}]
    return [{"page": 1, "text": path.read_text(encoding="utf-8", errors="ignore")}]



# ---------------------------------------------------------------------------
# The 7 chunking strategies — each returns a flat list of chunk strings
# ---------------------------------------------------------------------------

def chunk_fixed(text, chunk_size_chars, overlap_chars):
    splitter = CharacterTextSplitter(
        separator="\n", chunk_size=chunk_size_chars, chunk_overlap=overlap_chars
    )
    return splitter.split_text(text)


def chunk_recursive(text, chunk_size_chars, overlap_chars):
    splitter = RecursiveCharacterTextSplitter(
        separators=["\n\n", "\n", ". ", " ", ""],
        chunk_size=chunk_size_chars, chunk_overlap=overlap_chars
    )
    return splitter.split_text(text)


def chunk_sliding_window(text, chunk_size_chars, overlap_chars):
    step = max(chunk_size_chars - overlap_chars, 1)
    end = max(len(text) - chunk_size_chars + 1, 1)
    return [text[i:i + chunk_size_chars] for i in range(0, end, step)]

MIN_CHUNK_CHARS = 40  # drop/merge fragments shorter than this

def clean_text(text):
    """Strip PDF TOC/dot-leader junk and bare page numbers before splitting."""
    text = re.sub(r"\.{3,}", " ", text)          # dot leaders: "..... 10"
    text = re.sub(r"^\s*\d+\s*$", "", text, flags=re.MULTILINE)  # bare page numbers
    text = re.sub(r"[ \t]+", " ", text)
    return text

def chunk_semantic(text, chunk_size_chars, overlap_chars, similarity_threshold=0.5):
    if not SEMANTIC_AVAILABLE:
        return []
    text = clean_text(text)
    sentences = [s.strip() for s in re.split(r"(?<=[.!?])\s+", text.strip()) if s.strip()]
    if len(sentences) < 2:
        return sentences

    model = get_model()
    sentence_embeddings = model.encode(sentences)

    chunks, current = [], [sentences[0]]
    for i in range(1, len(sentences)):
        a, b = sentence_embeddings[i - 1], sentence_embeddings[i]
        sim = float(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b) + 1e-9))
        candidate_len = sum(len(s) for s in current) + len(sentences[i])
        if sim < similarity_threshold or candidate_len > chunk_size_chars:
            if sum(len(s) for s in current) < MIN_CHUNK_CHARS:
                current.append(sentences[i])
                continue
            chunks.append(" ".join(current))
            current = [sentences[i]]
        else:
            current.append(sentences[i])
    if current:
        chunks.append(" ".join(current))

    chunks = [c for c in chunks if len(c) >= MIN_CHUNK_CHARS]

    return chunks


def chunk_parent_child(text, chunk_size_chars, overlap_chars):
    # Compared here on the child (retrieval) unit, since that's what gets
    # matched against queries — the parent is only relevant once you wire
    # this into actual retrieval, not for structural comparison.
    parent_splitter = RecursiveCharacterTextSplitter(
        separators=["\n\n", "\n", ". ", " ", ""],
        chunk_size=chunk_size_chars * 4, chunk_overlap=overlap_chars
    )
    child_splitter = RecursiveCharacterTextSplitter(
        separators=["\n\n", "\n", ". ", " ", ""],
        chunk_size=chunk_size_chars, chunk_overlap=max(overlap_chars // 2, 0)
    )
    children = []
    for parent_text in parent_splitter.split_text(text):
        children.extend(child_splitter.split_text(parent_text))
    return children


def chunk_markdown_aware(text, chunk_size_chars, overlap_chars):
    headers = [("#", "Header 1"), ("##", "Header 2"), ("###", "Header 3")]
    header_splitter = MarkdownHeaderTextSplitter(headers_to_split_on=headers, strip_headers=False)
    sections = header_splitter.split_text(text)
    recursive = RecursiveCharacterTextSplitter(separators=["\n\n", "\n", ". ", " ", ""], chunk_size=chunk_size_chars, chunk_overlap=overlap_chars)
    out = []
    for s in sections:
        if len(s.page_content) <= chunk_size_chars:
            out.append(s.page_content)
        else:
            out.extend(recursive.split_text(s.page_content))
    return out


def chunk_html_aware(text, chunk_size_chars, overlap_chars):
    headers = [("h1", "Header 1"), ("h2", "Header 2"), ("h3", "Header 3")]
    splitter = HTMLHeaderTextSplitter(headers_to_split_on=headers)
    sections = splitter.split_text(text)
    recursive = RecursiveCharacterTextSplitter(separators=["\n\n", "\n", ". ", " ", ""], chunk_size=chunk_size_chars, chunk_overlap=overlap_chars)
    out = []
    for s in sections:
        if len(s.page_content) <= chunk_size_chars:
            out.append(s.page_content)
        else:
            out.extend(recursive.split_text(s.page_content))
    return out


METHODS = {
    # "fixed":         {"fn": chunk_fixed,         "applies_to": {".pdf", ".txt", ".md"}},
    "recursive":     {"fn": chunk_recursive,     "applies_to": {".pdf", ".txt", ".md"}},
    "sliding_window":{"fn": chunk_sliding_window,"applies_to": {".pdf", ".txt", ".md"}},
    "semantic":      {"fn": chunk_semantic,      "applies_to": {".pdf", ".txt", ".md"}},
    "parent_child":  {"fn": chunk_parent_child,  "applies_to": {".pdf", ".txt", ".md"}},
    "markdown_aware":{"fn": chunk_markdown_aware,"applies_to": {".md"}},
    # "html_aware":    {"fn": chunk_html_aware,    "applies_to": {".html", ".htm"}},
}


# ---------------------------------------------------------------------------
# Metrics
# ---------------------------------------------------------------------------

_SENTENCE_END_RE = re.compile(r'[.!?]["\')\]]?\s*$')


def is_mid_sentence_cut(chunk: str) -> bool:
    chunk = chunk.strip()
    if not chunk:
        return False
    return not bool(_SENTENCE_END_RE.search(chunk))


def compute_metrics(chunks, elapsed_sec):
    if not chunks:
        return dict(chunk_count=0, avg_len=0, median_len=0, min_len=0, max_len=0,
                     std_len=0, pct_mid_sentence_cut=0, elapsed_sec=round(elapsed_sec, 4))
    lengths = pd.Series([len(c) for c in chunks])
    mid_cuts = sum(is_mid_sentence_cut(c) for c in chunks)
    return dict(
        chunk_count=len(chunks),
        avg_len=round(lengths.mean(), 1),
        median_len=round(lengths.median(), 1),
        min_len=int(lengths.min()),
        max_len=int(lengths.max()),
        std_len=round(lengths.std() or 0, 1),
        pct_mid_sentence_cut=round(100 * mid_cuts / len(chunks), 1),
        elapsed_sec=round(elapsed_sec, 4),
    )


# ---------------------------------------------------------------------------
# Harness
# ---------------------------------------------------------------------------

def run_harness(paths, chunk_size_tokens_list=(256, 512), overlap_pct=15):
    rows = []
    all_chunks = []  # flat list of chunk records for the JSON export
    chunk_counter = 0

    for path in paths:
        path = Path(path)
        print(f"Processing {path.name} ...")
        pages = extract_text(path)
        suffix = path.suffix.lower()

        for target_tokens in chunk_size_tokens_list:
            chunk_size_chars = tokens_to_chars(target_tokens)
            overlap_chars = chunk_size_chars * overlap_pct // 100

            for method_name, spec in METHODS.items():
                if suffix not in spec["applies_to"]:
                    continue

                start = time.perf_counter()
                method_chunks = []  # (chunk_text, page) pairs for this method/size
                try:
                    for page_entry in pages:
                        page_chunks = spec["fn"](page_entry["text"], chunk_size_chars, overlap_chars)
                        # chunk_semantic can return (chunks, embeddings) — unwrap if so
                        if isinstance(page_chunks, tuple):
                            page_chunks = page_chunks[0]
                        method_chunks.extend((c, page_entry["page"]) for c in page_chunks)
                except Exception as e:
                    print(f"  [{method_name} @ {target_tokens}tok] FAILED: {e}")
                    continue
                elapsed = time.perf_counter() - start

                just_text = [c for c, _ in method_chunks]
                metrics = compute_metrics(just_text, elapsed)
                metrics.update(document=path.name, method=method_name, target_tokens=target_tokens)
                rows.append(metrics)

                for chunk_text, page_num in method_chunks:
                    all_chunks.append({
                        "chunk_id": f"chunk_{chunk_counter}",
                        "text": chunk_text,
                        "source": path.name,
                        "page": page_num,
                        "method": method_name,
                        "target_tokens": target_tokens,
                    })
                    chunk_counter += 1

    df = pd.DataFrame(rows)
    cols = ["document", "method", "target_tokens", "chunk_count", "avg_len",
            "median_len", "min_len", "max_len", "std_len",
            "pct_mid_sentence_cut", "elapsed_sec"]
    df = df[cols] if not df.empty else df
    return df, all_chunks


def print_summary(df: pd.DataFrame):
    if df.empty:
        print("No results — check your inputs.")
        return

    print("\n" + "=" * 100)
    print("FULL RESULTS (per document x method x chunk size)")
    print("=" * 100)
    print(df.to_string(index=False))

    print("\n" + "=" * 100)
    print("AVERAGED ACROSS ALL DOCUMENTS (per method x chunk size)")
    print("=" * 100)
    summary = (
        df.groupby(["method", "target_tokens"])
        .agg(
            avg_chunk_count=("chunk_count", "mean"),
            avg_len=("avg_len", "mean"),
            avg_std_len=("std_len", "mean"),
            avg_pct_mid_sentence_cut=("pct_mid_sentence_cut", "mean"),
            avg_elapsed_sec=("elapsed_sec", "mean"),
        )
        .round(2)
        .reset_index()
        .sort_values(["target_tokens", "avg_pct_mid_sentence_cut"])
    )
    print(summary.to_string(index=False))
    print(
        "\nReminder: pct_mid_sentence_cut is a STRUCTURAL proxy, not a retrieval-quality\n"
        "score. A method can win here and still lose on IoU/recall if it groups the\n"
        "wrong sentences together. Use this table to catch broken/degenerate splitters\n"
        "fast, then run the winners through chunking_evaluation (IoU/recall) before\n"
        "picking a default."
    )


# ---------------------------------------------------------------------------
# CONFIG — edit directly and re-run the cell/script. No argparse/CLI args,
# so this works the same whether you `python chunking_harness.py` locally
# or paste the whole file into a Colab cell and hit Run (Colab's own kernel
# args would break argparse, so we don't use it).
# ---------------------------------------------------------------------------

DOCS = [
    # add your own paths here, e.g. "/content/sample_docs/my_report.pdf"
]
CHUNK_SIZES_TOKENS = [512]
OVERLAP_PCT = 15
OUTPUT_CSV = "comparison_results.csv"

# Optional: point this at a folder instead of listing files individually.
# Auto-defaults to "sample_docs" (or "/content/sample_docs" on Colab) if that
# folder exists, so this works out of the box right after the download cell.
DOCS_DIR = None  # override e.g. "/content/sample_docs"

if DOCS_DIR is None:
    for _candidate in ("/content/sample_docs", "sample_docs"):
        if Path(_candidate).is_dir():
            DOCS_DIR = _candidate
            break


def _resolve_docs():
    paths = list(DOCS)
    if DOCS_DIR:
        supported = {".pdf", ".txt", ".md", ".html", ".htm"}
        paths += sorted(str(p) for p in Path(DOCS_DIR).iterdir() if p.suffix.lower() in supported)
    return paths


if __name__ == "__main__":
    import json

    input_paths = _resolve_docs()
    if not input_paths:
        raise SystemExit(
            "No documents found. Either:\n"
            "  1) Run the download cell first so ./sample_docs exists (auto-detected), or\n"
            "  2) Edit the DOCS list or DOCS_DIR near the bottom of this script, then re-run this cell."
        )
    results_df, all_chunks = run_harness(input_paths, chunk_size_tokens_list=CHUNK_SIZES_TOKENS, overlap_pct=OVERLAP_PCT)
    print_summary(results_df)

    summary_df = (
        results_df.groupby(["method", "target_tokens"])
        .agg(
            avg_chunk_count=("chunk_count", "mean"),
            avg_len=("avg_len", "mean"),
            avg_std_len=("std_len", "mean"),
            avg_pct_mid_sentence_cut=("pct_mid_sentence_cut", "mean"),
            avg_elapsed_sec=("elapsed_sec", "mean"),
        )
        .round(2)
        .reset_index()
    )

    excel_path = OUTPUT_CSV.replace(".csv", ".xlsx")
    with pd.ExcelWriter(excel_path) as writer:
        results_df.to_excel(writer, sheet_name="full_results", index=False)
        summary_df.to_excel(writer, sheet_name="summary", index=False)
    print(f"\nSaved summary workbook to {excel_path}")

    chunks_path = OUTPUT_CSV.replace(".csv", "_chunks.json")
    with open(chunks_path, "w", encoding="utf-8") as f:
        json.dump(all_chunks, f, ensure_ascii=False, indent=2)
    print(f"Saved {len(all_chunks)} chunks to {chunks_path}")

Processing attention_is_all_you_need.pdf ...


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Processing langchain_readme.md ...
Processing nist_800-61r2.pdf ...
Processing pride_and_prejudice.txt ...

FULL RESULTS (per document x method x chunk size)
                     document         method  target_tokens  chunk_count  avg_len  median_len  min_len  max_len  std_len  pct_mid_sentence_cut  elapsed_sec
attention_is_all_you_need.pdf      recursive            512           27   1584.2      1653.0      740     2042    468.2                  92.6       0.0013
attention_is_all_you_need.pdf sliding_window            512           16   1802.8      2048.0      812     2048    493.2                 100.0       0.0000
attention_is_all_you_need.pdf       semantic            512          240    162.7       108.0       40      909    144.2                   1.2      12.0412
attention_is_all_you_need.pdf   parent_child            512           27   1512.7      1554.0      382     2043    525.6                  92.6       0.0021
          langchain_readme.md      recursive            512   

In [9]:
import json
import pandas as pd

with open("comparison_results_chunks.json", "r", encoding="utf-8") as f:
    raw_chunks = json.load(f)

chunks_df = pd.DataFrame(raw_chunks)
# existing cells expect a "title" and "doc_id" column — derive them from source
chunks_df["doc_id"] = chunks_df["source"]
chunks_df["title"] = chunks_df["source"].apply(lambda s: Path(s).stem)

print(chunks_df["method"].value_counts())
print(chunks_df["target_tokens"].value_counts())

method
semantic          5765
parent_child       679
recursive          659
sliding_window     549
markdown_aware       5
Name: count, dtype: int64
target_tokens
512    7657
Name: count, dtype: int64


In [10]:
def filter_chunks(df, method=None, target_tokens=None, source=None):
    """Filter chunks_df by any combination of metadata fields before indexing."""
    out = df
    if method is not None:
        out = out[out["method"] == method]
    if target_tokens is not None:
        out = out[out["target_tokens"] == target_tokens]
    if source is not None:
        out = out[out["source"] == source]
    return out.reset_index(drop=True)

# Set this after reading your Excel summary sheet and picking the winner
# (lowest pct_mid_sentence_cut / best std_len tradeoff for your data).
CHOSEN_METHOD = "semantic"
CHOSEN_TOKENS = 512

chunks_df = filter_chunks(chunks_df, method=CHOSEN_METHOD, target_tokens=CHOSEN_TOKENS)
chunks_df = chunks_df.reset_index(drop=True)

corpus = chunks_df["text"].tolist()
source = chunks_df["doc_id"].tolist()

print(f"Indexing {len(corpus)} chunks from method={CHOSEN_METHOD}, size={CHOSEN_TOKENS}")

Indexing 5765 chunks from method=semantic, size=512


In [11]:
_MODEL = None


def get_model():
    global _MODEL
    if _MODEL is None:
        _MODEL = SentenceTransformer("all-MiniLM-L6-v2")
    return _MODEL


model = get_model()
embedding = np.array(model.encode(corpus))
assert len(corpus) == len(embedding) == len(source)

client = chromadb.PersistentClient(path="./chroma_db2")
collection = client.get_or_create_collection(name="my_docs")

batch_size = 5000
for i in range(0, len(corpus), batch_size):
    end = min(i + batch_size, len(corpus))
    collection.add(
        documents=corpus[i:end],
        embeddings=embedding[i:end].tolist(),
        ids=[f"chunk_{j}" for j in range(i, end)],
        metadatas=[
            {
                "doc_id": str(source[j]),
                "title": str(chunks_df["title"].iloc[j]),
                "position": int(j),
                "page": int(chunks_df["page"].iloc[j]),
                "method": str(chunks_df["method"].iloc[j]),
            }
            for j in range(i, end)
        ],
    )

In [12]:
def get_bm25(corpus, stop_words=[]):
    tokens = bm25s.tokenize(corpus, stopwords=stop_words)
    bm25_retriever = bm25s.BM25()
    bm25_retriever.index(tokens, corpus)
    return bm25_retriever


bm25_index = get_bm25(corpus=corpus, stop_words="en")

Split strings:   0%|          | 0/5765 [00:00<?, ?it/s]

DEBUG:bm25s:Building index from IDs objects


BM25S Count Tokens:   0%|          | 0/5765 [00:00<?, ?it/s]

BM25S Compute Scores:   0%|          | 0/5765 [00:00<?, ?it/s]

In [13]:
def search_dense(collection, query: str, k: int = 10):
    result = collection.query(query_texts=[query], n_results=k)
    ids = [int(doc_id.split("_")[1]) for doc_id in result["ids"][0]]
    distances = result["distances"][0]
    return list(zip(ids, distances)), result


def search_bm25(query: str, k: int = 10):
    query_tokens = bm25s.tokenize([query], stopwords="en")
    indices, scores = bm25_index.retrieve(query_tokens, k=k)
    return indices, scores


K_RRF = 60


def reciprocal_rank_fusion(rankings: list, k: int = K_RRF):
    scores: dict = defaultdict(float)
    for ranking in rankings:
        for rank, doc_id in enumerate(ranking, start=1):
            scores[doc_id] += 1.0 / (k + rank)
    return sorted(scores.items(), key=lambda x: -x[1])


def search_hybrid(collection, query: str, k: int = 10, candidate_k: int = 10):
    bm25_raw_indices, _ = search_bm25(query, k=candidate_k)
    bm25_doc_ids = [str(doc_id) for doc_id in bm25_raw_indices[0]] if bm25_raw_indices.size > 0 else []

    dense_raw_results, _ = search_dense(collection, query, k=candidate_k)
    dense_doc_ids = [str(doc_id) for doc_id, _ in dense_raw_results]

    fused_results = reciprocal_rank_fusion([bm25_doc_ids, dense_doc_ids])
    return fused_results[:k], dense_doc_ids, bm25_doc_ids


def normalize_retrieval(method_name, query, k_candidates=20):
    if method_name == "dense":
        raw, _ = search_dense(collection, query, k=k_candidates)
        return [(int(pid), score) for pid, score in raw]
    if method_name == "bm25":
        indices, scores = search_bm25(query, k=k_candidates)
        return [(int(pid), float(s)) for pid, s in zip(indices[0], scores[0])]
    if method_name == "hybrid":
        fused, _, _ = search_hybrid(collection, query, k=k_candidates, candidate_k=k_candidates)
        return [(int(pid), score) for pid, score in fused]
    raise ValueError(f"unknown method: {method_name}")


def position_to_doc_id(position: int) -> str:
    return source[position]


In [14]:
from langchain_core.documents import Document
from ragas.testset import TestsetGenerator
from ragas.testset.persona import Persona
from ragas.testset.transforms import (
    KeyphrasesExtractor,
    OverlapScoreBuilder,
)
from ragas.testset.transforms.extractors.llm_based import NERExtractor


In [ ]:
GOOGLE_API_KEY = ""  # move to os.environ / Colab secret when you can
os.environ["GOOGLE_API_KEY"] = GOOGLE_API_KEY

gemini_client = genai.Client(api_key=GOOGLE_API_KEY)

evaluator_llm = LangchainLLMWrapper(
    ChatGoogleGenerativeAI(model="gemini-2.5-flash-lite", temperature=0)
)


/tmp/ipykernel_689/192574391.py:6: DeprecationWarning: LangchainLLMWrapper is deprecated and will be removed in a future version. Use llm_factory instead: from openai import OpenAI; from ragas.llms import llm_factory; llm = llm_factory('gpt-4o-mini', client=OpenAI(api_key='...'))
  evaluator_llm = LangchainLLMWrapper(


In [32]:
gen_llm = evaluator_llm

In [33]:
ner_extractor = NERExtractor(llm=gen_llm)
keyphrase_extractor = KeyphrasesExtractor(llm=gen_llm, property_name="keyphrases", max_num=10)
relation_builder = OverlapScoreBuilder(
    property_name="keyphrases",
    new_property_name="overlap_score",
    threshold=0.01,
    distance_threshold=0.9,
)

transforms = [ner_extractor, keyphrase_extractor, relation_builder]

In [34]:
from ragas.embeddings import LangchainEmbeddingsWrapper
from langchain_core.embeddings import Embeddings

class LocalSentenceTransformerEmbeddings(Embeddings):
    def __init__(self, st_model):
        self.st_model = st_model  # renamed from self.model to avoid collision
                                    # with ragas's usage-tracking, which expects
                                    # a `.model` attribute to be a string name

    def embed_documents(self, texts):
        return self.st_model.encode(texts).tolist()

    def embed_query(self, text):
        return self.st_model.encode([text])[0].tolist()

local_embeddings = LangchainEmbeddingsWrapper(LocalSentenceTransformerEmbeddings(model))

/tmp/ipykernel_689/1975738439.py:16: DeprecationWarning: LangchainEmbeddingsWrapper is deprecated and will be removed in a future version. Use the modern embedding providers instead: embedding_factory('openai', model='text-embedding-3-small', client=openai_client) or from ragas.embeddings import OpenAIEmbeddings, GoogleEmbeddings, HuggingFaceEmbeddings
  local_embeddings = LangchainEmbeddingsWrapper(LocalSentenceTransformerEmbeddings(model))


In [35]:
import os

os.environ["DEEPEVAL_PER_TASK_TIMEOUT_SECONDS_OVERRIDE"] = "600"
os.environ["DEEPEVAL_PER_ATTEMPT_TIMEOUT_SECONDS_OVERRIDE"] = "300"
os.environ["DEEPEVAL_TASK_GATHER_BUFFER_SECONDS_OVERRIDE"] = "60"

In [ ]:
personas = [
    Persona(
        name="ML researcher",
        role_description="A machine learning researcher looking for precise technical details about model architecture, training procedures, and benchmark results.",
    ),
    Persona(
        name="security analyst",
        role_description="A security professional trying to understand incident response procedures, roles, and compliance requirements.",
    ),
    Persona(
        name="software developer",
        role_description="A developer trying to understand how to use a software library or framework in their own project.",
    ),
    Persona(
        name="curious reader",
        role_description="A general reader interested in plot, character relationships, and events in a novel.",
    ),
]

generator = TestsetGenerator(
    llm=gen_llm,
    embedding_model=local_embeddings,
    persona_list=personas,
)

# ---------------------------------------------------------------------------
# Per-source generation: 5 questions per document, generated and saved
# one source at a time (not all 40 in one go) so progress isn't lost if
# a later source fails or the runtime disconnects.
# ---------------------------------------------------------------------------
import time
import pandas as pd

SYNTH_PER_SOURCE = 5
OUT_XLSX = "synthetic_testset.xlsx"

unique_sources = chunks_df["doc_id"].unique().tolist()
all_synth_rows = []

for src in unique_sources:
    print(f"Generating {SYNTH_PER_SOURCE} synthetic questions for: {src}")

    src_docs = [
        Document(
            page_content=row["text"],
            metadata={"doc_id": row["doc_id"], "title": row["title"], "page": row["page"]},
        )
        for _, row in chunks_df[chunks_df["doc_id"] == src].iterrows()
    ]

    try:
        src_dataset = generator.generate_with_langchain_docs(
            src_docs,
            testset_size=SYNTH_PER_SOURCE,
            transforms=transforms,
        )
    except Exception as e:
        print(f"  FAILED for {src}: {e}")
        continue

    src_df = src_dataset.to_pandas()
    src_df["source"] = src
    all_synth_rows.append(src_df)

    # save after every source, not just at the end
    combined_so_far = pd.concat(all_synth_rows, ignore_index=True)
    combined_so_far.to_excel(OUT_XLSX, index=False)
    print(f"  Saved {len(combined_so_far)} total rows to {OUT_XLSX}")

    time.sleep(1)  # small buffer between sources for API rate limits

synthetic_df = pd.concat(all_synth_rows, ignore_index=True)
print(f"\nDone. {len(synthetic_df)} synthetic rows across {len(unique_sources)} sources.")

In [37]:
import pandas as pd

synthetic_df = pd.read_excel("synthetic_testset.xlsx")
print(f"Loaded {len(synthetic_df)} rows from synthetic_testset.xlsx")
print(synthetic_df.columns.tolist())

# Build the (qid, query_text, ground_truth) rows list that build_dataset_for_config expects
rows = [
    (f"synth_{i}", row["user_input"], row["reference"])
    for i, row in synthetic_df.iterrows()
    if pd.notna(row.get("user_input")) and pd.notna(row.get("reference"))
]

print(f"Prepared {len(rows)} rows for evaluation (out of {len(synthetic_df)} total synthetic rows)")

Loaded 10 rows from synthetic_testset.xlsx
['user_input', 'reference_contexts', 'reference', 'persona_name', 'query_style', 'query_length', 'synthesizer_name', 'source']
Prepared 10 rows for evaluation (out of 10 total synthetic rows)


In [38]:
_RERANKER = None


def get_reranker():
    global _RERANKER
    if _RERANKER is None:
        _RERANKER = FlagReranker("BAAI/bge-reranker-v2-m3", use_fp16=True)
    return _RERANKER


def reranking(query, chunks, k=None):
    """chunks: list of [text, id]. Returns [[score, id], ...] sorted desc."""
    reranker = get_reranker()
    pairs = [[query, text] for text, _id in chunks]
    scores = reranker.compute_score(pairs)
    if isinstance(scores, float):
        scores = [scores]
    result = sorted(
        ([score, _id] for score, (_, _id) in zip(scores, chunks)),
        key=lambda x: x[0],
        reverse=True,
    )
    return result[:k] if k else result


def rerank_retrieval(method_name, query, k=10, k_candidates=20, rerank_fn=reranking):
    before = normalize_retrieval(method_name, query, k_candidates=k_candidates)
    chunk_pairs = [[corpus[pos], pos] for pos, _ in before]
    reranked = rerank_fn(query, chunk_pairs)
    return {
        "before_ids": [pos for pos, _ in before][:k],
        "before_scores": [s for _, s in before][:k],
        "after_ids": [pos for _, pos in reranked][:k],
        "after_scores": [s for s, _ in reranked][:k],
    }


In [39]:
# id -> metadata lookup, built once, used to filter retrieval results before generation
position_to_meta = {
    j: {
        "doc_id": source[j],
        "page": chunks_df["page"].iloc[j],
        "method": chunks_df["method"].iloc[j],
    }
    for j in range(len(corpus))
}

def filter_positions_by_metadata(positions, source_filter=None, page_range=None):
    """Drop retrieved chunk positions that don't match the given metadata constraints."""
    filtered = []
    for pos in positions:
        meta = position_to_meta.get(pos, {})
        if source_filter is not None and meta.get("doc_id") != source_filter:
            continue
        if page_range is not None:
            page = meta.get("page")
            if page is None or not (page_range[0] <= page <= page_range[1]):
                continue
        filtered.append(pos)
    return filtered


def retrieve_contexts_config(query: str, method: str, use_rerank: bool, top_k: int = 5,
                              k_candidates: int = 20, source_filter=None, page_range=None):
    """Same as before, with an added metadata filter step before contexts are built."""
    t0 = time.perf_counter()

    if use_rerank:
        r = rerank_retrieval(method, query, k=k_candidates, k_candidates=k_candidates, rerank_fn=reranking)
        positions = r["after_ids"]
    else:
        raw = normalize_retrieval(method, query, k_candidates=k_candidates)
        positions = [pos for pos, _ in raw]

    positions = filter_positions_by_metadata(positions, source_filter=source_filter, page_range=page_range)
    positions = positions[:top_k]

    latency = time.perf_counter() - t0
    contexts = [corpus[pos] for pos in positions]
    return contexts, latency

In [40]:
def generate_answer(query: str, contexts: list) -> str:
    context_text = "\n\n".join(contexts)
    prompt = f"""Answer the question directly and concisely, using ONLY the context below.
Do not comment on whether the context is complete or sufficient — just answer with whatever relevant information is present.
Only if the context contains genuinely nothing relevant to the question, respond with exactly: "I don't know based on the provided context."

Context:
{context_text}

Question: {query}

Answer:"""
    response = gemini_client.models.generate_content(
        model="gemini-2.5-flash-lite",
        contents=prompt,
    )
    return response.text.strip()


In [41]:
CONFIGS = [
    ("bm25", False), ("bm25", True),
    ("dense", False), ("dense", True),
    ("hybrid", False), ("hybrid", True),
]

def build_dataset_for_config(method, use_rerank, rows):
    data = {"question": [], "answer": [], "contexts": [], "ground_truth": []}
    latencies = []
    for qid, query_text, gt in rows:  # your merged 40-60 row set
        contexts, lat = retrieve_contexts_config(query_text, method, use_rerank)
        answer = generate_answer(query_text, contexts)
        data["question"].append(query_text)
        data["answer"].append(answer)
        data["contexts"].append(contexts)
        data["ground_truth"].append(gt)
        latencies.append(lat)
    return Dataset.from_dict(data), sum(latencies) / len(latencies)

In [ ]:
from ragas import evaluate
from ragas.metrics import faithfulness, answer_relevancy, context_precision, context_recall


In [44]:
import time

results_table = []
per_config_datasets = {}

for method, use_rerank in CONFIGS:
    print(f"Running config: method={method}, rerank={use_rerank}")

    try:
        dataset, avg_latency = build_dataset_for_config(method, use_rerank, rows)

        result = evaluate(
            dataset,
            metrics=[faithfulness, answer_relevancy, context_precision, context_recall],
            llm=evaluator_llm,
            embeddings=local_embeddings,
        )

        result_df = result.to_pandas()
        result_df["method"] = method
        result_df["rerank"] = use_rerank
        per_config_datasets[f"{method}_{use_rerank}"] = result_df

        row = result_df.mean(numeric_only=True).to_dict()
        row["method"] = method
        row["rerank"] = use_rerank
        row["avg_latency_ms"] = avg_latency * 1000
        results_table.append(row)

        # save after EVERY config, not just at the end
        pd.DataFrame(results_table).to_csv("final_eval_results_partial.csv", index=False)
        print(f"  Done. Saved partial results ({len(results_table)}/{len(CONFIGS)} configs)")

    except Exception as e:
        print(f"  FAILED config method={method}, rerank={use_rerank}: {e}")
        print("  Skipping — will not lose previously completed configs")
        continue

    time.sleep(5)  # brief pause between configs to ease quota pressure

results_df = pd.DataFrame(results_table)
with pd.ExcelWriter("final_eval_results.xlsx") as writer:
    results_df.to_excel(writer, sheet_name="summary", index=False)
    for config_name, df in per_config_datasets.items():
        df.to_excel(writer, sheet_name=config_name[:31], index=False)

print("Saved final_eval_results.xlsx")
results_df

Running config: method=bm25, rerank=False


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Evaluating:   0%|          | 0/40 [00:00<?, ?it/s]

  Done. Saved partial results (1/6 configs)
Running config: method=bm25, rerank=True


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Evaluating:   0%|          | 0/40 [00:00<?, ?it/s]

  Done. Saved partial results (2/6 configs)
Running config: method=dense, rerank=False


/root/.cache/chroma/onnx_models/all-MiniLM-L6-v2/onnx.tar.gz: 100%|██████████| 79.3M/79.3M [00:01<00:00, 46.6MiB/s]


Evaluating:   0%|          | 0/40 [00:00<?, ?it/s]

  Done. Saved partial results (3/6 configs)
Running config: method=dense, rerank=True


Evaluating:   0%|          | 0/40 [00:00<?, ?it/s]

  Done. Saved partial results (4/6 configs)
Running config: method=hybrid, rerank=False


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Evaluating:   0%|          | 0/40 [00:00<?, ?it/s]

  Done. Saved partial results (5/6 configs)
Running config: method=hybrid, rerank=True


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

Evaluating:   0%|          | 0/40 [00:00<?, ?it/s]

  Done. Saved partial results (6/6 configs)
Saved final_eval_results.xlsx


,faithfulness,answer_relevancy,context_precision,context_recall,rerank,method,avg_latency_ms
0,1.000000,0.477670,0.748333,0.9,False,bm25,36.924681
1,0.983333,0.478213,0.875000,1.0,True,bm25,498.252199
2,0.900000,0.352699,0.509306,0.7,False,dense,723.326645
3,1.000000,0.426748,0.651667,0.8,True,dense,831.096150
4,1.000000,0.350967,0.615000,0.8,False,hybrid,425.073469
5,0.983333,0.468065,0.870000,1.0,True,hybrid,923.096092


## Key Findings (Month 1 — Advanced RAG)

Evaluated on 10 hand-written queries across bm25/dense/hybrid retrieval,
with and without BGE reranking (bge-reranker-v2-m3), using RAGAS metrics.

| Method | Rerank | Faithfulness | Answer Relevancy | Context Precision | Context Recall |
|---|---|---|---|---|---|
| bm25    | No  | 1.000 | 0.478 | 0.748 | 0.90 |
| bm25    | Yes | 0.983 | 0.478 | 0.875 | 1.00 |
| dense   | No  | 0.900 | 0.353 | 0.509 | 0.70 |
| dense   | Yes | 1.000 | 0.427 | 0.652 | 0.80 |
| hybrid  | No  | 1.000 | 0.351 | 0.615 | 0.80 |
| hybrid  | Yes | 0.983 | 0.468 | 0.870 | 1.00 |

**Takeaway:** Reranking improved context precision and recall in every
configuration tested, most notably for bm25 and hybrid retrieval
(precision +0.13–0.25, recall to 1.0). Dense-only retrieval was the
weakest baseline. bm25+rerank and hybrid+rerank were roughly tied as
the best-performing configs on this query set.

**Limitation:** Evaluated on 10 queries due to Gemini free-tier API
quota constraints during the run. Results are directionally consistent
but a larger query set (40-60, as originally planned) would be needed
to confirm statistical significance. Wider validation is noted as
future work.